From 4413523b6734f4e745a425ef14d7aa0c2624f0ce Mon Sep 17 00:00:00 2001
From: Limeng <Meng.Li@windriver.com>
Date: Wed, 5 Apr 2017 18:36:38 +0800
Subject: [PATCH 2/3] arch: arm: mach-owl: remove the improper invocation of
 clk_put() in owl_i2c_set_clk_tree

detail crash info as below:
Unable to handle kernel paging request at virtual address 2d6332a5
pgd = c0004000
[2d6332a5] *pgd=00000000
Internal error: Oops: 5 [#1] PREEMPT SMP ARM
Modules linked in: pvrsrvkm drm atc260x_irkeypad sch_fq_codel softdog
CPU: 0 PID: 61 Comm: kworker/u8:1 Not tainted 4.1.21-WR8.0.0.0_standard #26
Hardware name: gs705a
Workqueue: suspend early_suspend
task: eb7d4ec0 ti: ea8f0000 task.ti: ea8f0000
PC is at clk_core_enable+0x1c/0x230
LR is at clk_enable+0x30/0x44
pc : [<c05c75d8>]    lr : [<c05c781c>]    psr: 200001d3
sp : ea8f1d68  ip : ea8f1d88  fp : ea8f1d84
r10: c0c0ef60  r9 : eb615010  r8 : c09ca276
r7 : 00000010  r6 : eb621a3c  r5 : 60000153  r4 : 2d633269
r3 : 00000001  r2 : 00000000  r1 : 00ed00ec  r0 : 2d633269
Flags: nzCv  IRQs off  FIQs off  Mode SVC_32  ISA ARM  Segment kernel
Control: 10c5387d  Table: 2899404a  DAC: 00000015
Process kworker/u8:1 (pid: 61, stack limit = 0xea8f0210)
Stack: (0xea8f1d68 to 0xea8f2000)
1d60:                   eb5b5840 60000153 eb621a3c 00000010 ea8f1d9c ea8f1d88
1d80: c05c781c c05c75c8 eb621800 c00253f4 ea8f1dbc ea8f1da0 c0025430 c05c77f8
1da0: eb615010 c00253f4 eb615010 00000010 ea8f1dec ea8f1dc0 c0439448 c0025400
1dc0: ea8f1dec ea8f1dd0 eb615010 00000010 00000010 c0c0ef60 c0c43328 eb615010
1de0: ea8f1e04 ea8f1df0 c04396a8 c0439394 eb615078 c0cdf518 ea8f1e4c ea8f1e08
1e00: c043a7fc c043958c c0c42e78 00000000 9588bf33 0000000d 9588bf33 0000000d
1e20: c0bba174 00000000 00000003 c0c4a6bc 00000000 c0c43328 c0c42e78 00000000
1e40: ea8f1e94 ea8f1e50 c0072fb0 c043a684 c0bc2ac8 ea8f1e74 c0c43328 00000000
1e60: ea8f1e94 ea8f1e80 c0072b2c 00000000 c0c43328 00000003 c0bc2ac8 c0bc2ae0
1e80: c0c42e78 00000000 ea8f1ebc ea8f1e98 c0073c50 c0072dd0 c0bc2ae0 c0c4a8b8
1ea0: c0bc2ac8 c0bc2ac8 c0bc2ae0 c0c42e78 ea8f1ee4 ea8f1ec0 c007a394 c00736a8
1ec0: eb687600 c0bc2af0 eb436800 00000000 eb60a300 c0c42e78 ea8f1f24 ea8f1ee8
1ee0: c0049e78 c007a2d0 ea8f1f0c ea8f1ef8 c0058074 c0057f60 eb436800 eb687600
1f00: eb436800 eb436800 c0ba8a40 eb436814 eb687618 00000088 ea8f1f5c ea8f1f28
1f20: c004ac78 c0049c24 eb687600 c004a9ac 00000000 eb686300 00000000 eb687600
1f40: c004a9ac 00000000 00000000 00000000 ea8f1fac ea8f1f60 c004f770 c004a9b8
1f60: c080c404 00000000 ea8f1f94 eb687600 00000000 00000000 ea8f1f78 ea8f1f78
1f80: 00000000 00000000 ea8f1f88 ea8f1f88 eb686300 c004f688 00000000 00000000
1fa0: 00000000 ea8f1fb0 c000f3c8 c004f694 00000000 00000000 00000000 00000000
1fc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
1fe0: 00000000 00000000 00000000 00000000 00000013 00000000 ffffffff ffffffff
[<c05c75d8>] (clk_core_enable) from [<c05c781c>] (clk_enable+0x30/0x44)
[<c05c781c>] (clk_enable) from [<c0025430>] (owl_i2c_resume+0x3c/0x9c)
[<c0025430>] (owl_i2c_resume) from [<c0439448>] (dpm_run_callback+0xc0/0x1f8)
[<c0439448>] (dpm_run_callback) from [<c04396a8>] (device_resume_noirq+0x128/0x164)
[<c04396a8>] (device_resume_noirq) from [<c043a7fc>] (dpm_resume_noirq+0x184/0x3a4)
[<c043a7fc>] (dpm_resume_noirq) from [<c0072fb0>] (suspend_devices_and_enter+0x1ec/0x8d8)
[<c0072fb0>] (suspend_devices_and_enter) from [<c0073c50>] (pm_suspend+0x5b4/0x6bc)
[<c0073c50>] (pm_suspend) from [<c007a394>] (early_suspend+0xd0/0x118)
[<c007a394>] (early_suspend) from [<c0049e78>] (process_one_work+0x260/0x49c)
[<c0049e78>] (process_one_work) from [<c004ac78>] (worker_thread+0x2cc/0x40c)
[<c004ac78>] (worker_thread) from [<c004f770>] (kthread+0xe8/0xfc)
[<c004f770>] (kthread) from [<c000f3c8>] (ret_from_fork+0x14/0x2c)
Code: e52de004 e8bd4000 e2504000 0a000007 (e594303c)
---[ end trace 253c9d26675d84ab ]---

The reason is that i2c_clk is released when call clk_put() function during
probe stage. So, delete clk_put() function invoking.
This issue is introduced by commit b11912c6f323
(arm: owl: initial support Actions S500 SoC) when merge SDK patch related with
s500 mach.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-owl/i2c-owl.c |    4 ----
 1 files changed, 0 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-owl/i2c-owl.c b/arch/arm/mach-owl/i2c-owl.c
index 7b5d2ad..548c953 100755
--- a/arch/arm/mach-owl/i2c-owl.c
+++ b/arch/arm/mach-owl/i2c-owl.c
@@ -1250,7 +1250,6 @@ static int owl_i2c_set_clk_tree(struct platform_device *pdev)
 			goto round_rate_failed;
 		clk_prepare(dev->i2c_clk);
 		clk_enable(dev->i2c_clk);
-		clk_put(dev->i2c_clk);
 		module_clk_enable(MOD_ID_I2C0);
 		break;
 
@@ -1263,7 +1262,6 @@ static int owl_i2c_set_clk_tree(struct platform_device *pdev)
 			goto round_rate_failed;
 		clk_prepare(dev->i2c_clk);
 		clk_enable(dev->i2c_clk);
-		clk_put(dev->i2c_clk);
 		module_clk_enable(MOD_ID_I2C1);
 		break;
 
@@ -1276,7 +1274,6 @@ static int owl_i2c_set_clk_tree(struct platform_device *pdev)
 			goto round_rate_failed;
 		clk_prepare(dev->i2c_clk);
 		clk_enable(dev->i2c_clk);
-		clk_put(dev->i2c_clk);
 		module_clk_enable(MOD_ID_I2C2);
 		break;
 
@@ -1289,7 +1286,6 @@ static int owl_i2c_set_clk_tree(struct platform_device *pdev)
 			goto round_rate_failed;
 		clk_prepare(dev->i2c_clk);
 		clk_enable(dev->i2c_clk);
-		clk_put(dev->i2c_clk);
 		module_clk_enable(MOD_ID_I2C3);
 		break;
 
-- 
1.7.5.4

